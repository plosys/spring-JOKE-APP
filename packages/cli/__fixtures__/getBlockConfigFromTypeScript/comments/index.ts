import '@appsemble/sdk';

declare module '@appsemble/sdk' {
  interface Actions {
    /**
     * Valid action comment
     */
    comment: never;

    /**
     * Unexpected comment
     */
    /**
     * Expected comment
     */
    duplicate: never;

    // Ignored line comment
    line: never;
  }

  interface EventEmitters {
    /**
     * Test event emitter.
     */
    testEmit: never;
  }

  interface EventListeners {
    /**
     * Test event listener.
     */
    testListener: never;
  }

  // Comments are extracted by typescript-json-schema
  interface Parameters {
    param: any;
  }
}
