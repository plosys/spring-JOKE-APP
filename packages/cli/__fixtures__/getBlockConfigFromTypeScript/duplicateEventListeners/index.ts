import '@appsemble/sdk';

declare module '@appsemble/sdk' {
  interface Actions {
    testAction: never;
  }

  interface EventEmitters {
    testEmit: never;
  }

  interface EventListeners {
    testListener: never;
  }

  interface Parameters {
    param1: string;
    param2: number;
    param3: boolean;
    param4?: string;
    param5: {
      nested1: string;
      nested2: number;
      nested3: boolean;
    };
    param6: string[];
  }
}

declare module '@appsemble/sdk' {
  interface EventListeners {
    duplicate: never;
  }
}
