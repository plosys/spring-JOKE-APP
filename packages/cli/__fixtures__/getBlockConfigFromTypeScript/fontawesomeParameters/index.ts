import '@appsemble/sdk';
import { IconName } from '@fortawesome/fontawesome-common-types';

declare module '@appsemble/sdk' {
  interface Parameters {
    /**
     * This is an icon.
     */
    icon?: IconName;
  }
}
