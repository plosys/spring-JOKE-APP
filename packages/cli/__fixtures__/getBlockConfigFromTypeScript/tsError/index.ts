import '@appsemble/sdk';

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const unused = undefined;
