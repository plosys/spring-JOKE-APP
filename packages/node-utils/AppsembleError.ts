
/**
 * Used for throwing known Appsemble errors.
 *
 * This is used when the normal control flow needs to be aborted, but for reasons known. When an
 * instance of `AppsembleError` is thrown, the message will be logged, but not the stack trace.
 */
export class AppsembleError extends Error {
  /**
   * @param message The error message to show to the user.
   */
  constructor(message: string) {
    super(message);
    this.name = 'AppsembleError';
    Error.captureStackTrace(this, this.constructor);
  }
}