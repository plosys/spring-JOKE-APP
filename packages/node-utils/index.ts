
export * from './AppsembleError.js';
export * from './basicAuth.js';
export * from './formData.js';
export * from './fs.js';
export * from './getWorkspaces.js';
export * from './handleError.js';
export * from './interceptors.js';
export * from './logger.js';
export * from './loggerMiddleware.js';
export * from './readFileOrString.js';
export * from './testFixtures.js';