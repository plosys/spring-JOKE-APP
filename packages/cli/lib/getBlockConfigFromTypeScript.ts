
import { relative } from 'node:path';

import { AppsembleError, logger } from '@appsemble/node-utils';
import { BlockConfig, BlockManifest } from '@appsemble/types';
import { parse } from 'comment-parser';
import { Schema } from 'jsonschema';
import normalizePath from 'normalize-path';
import { createFormatter, createParser, SchemaGenerator, ts } from 'ts-json-schema-generator';

/**
 * Get the tsdoc comment for a TypeScript node.
 *
 * @param checker The type checker instance to use.
 * @param node The node for which to get the tsdoc.
 * @returns The tsdoc comment as a string, if present.
 */
function getNodeComments(checker: ts.TypeChecker, node: ts.TypeElement): string {
  const symbol = checker.getSymbolAtLocation(node.name);
  if (!symbol) {
    return;
  }

  const comments = symbol.getDocumentationComment(checker);
  if (comments.length) {
    return comments
      .map((comment) =>
        comment.kind === 'lineBreak' ? comment.text : comment.text.trim().replace(/\r\n/g, '\n'),
      )
      .join('');
  }
}

/**
 * Process keys from an interface into an object.
 *
 * This asserts only index signature members and regular alphanumerical properties are used.
 *
 * @param iface The TypeScript interface to check.
 * @param checker The TypeScript type checker.
 * @param convert A function for converting extracted data to a value. It will be called with the
 * name and JSDoc description. The name is `undefined` for index signatures. The function should
 * return a tuple of key and value.
 * @returns A record created from the returned key/value pairs.
 */
function processInterface<T>(
  iface: ts.InterfaceDeclaration,
  checker: ts.TypeChecker,
  convert: (name: string, description: string) => [string, T],
): Record<string, T> {
  if (!iface?.members.length) {
    return;
  }

  const source = iface.getSourceFile().getText();

  return Object.fromEntries(
    iface.members.map((member) => {
      const description = getNodeComments(checker, member);
      if (ts.isIndexSignatureDeclaration(member)) {
        // Comments aren’t properly extracted for index signatures.
        const commentRanges = ts.getLeadingCommentRanges(source, member.pos);
        if (commentRanges) {
          const comments = commentRanges?.map((r) => source.slice(r.pos, r.end)) ?? [];
          const [block] = parse(comments[0]);
          const processedDescription = block.source
            .map((src) => src.tokens.description)
            .join('\n')
            .replace(/^\s+|\s+$/g, '');

          return convert(undefined, processedDescription);
        }

        return convert(undefined, description);
      }

      if (!ts.isPropertySignature(member)) {
        throw new AppsembleError(
          `Only property and index signatures are allowed as ${
            iface.name
          } keys. Found: ${member.getFullText()}`,
        );
      }

      const { name } = member;

      if (!ts.isIdentifier(name)) {
        throw new AppsembleError(
          `Only property and index signatures are allowed as ${
            iface.name
          } keys. Found: ${name.getFullText()}`,
        );
      }

      const { text } = name;

      if (!/^[\da-z]+$/i.test(text)) {
        throw new AppsembleError(
          `Found property named ‘${text}’ in ${iface.name} interface. Only alphanumerical identifiers are supported.`,
        );
      }

      return convert(text, description);
    }),
  );
}

/**
 * Get an actions object based on a TypeScript interface node.
 *
 * @param iface The node to base the actions on.
 * @param checker The TypeScript type checker.
 * @returns The action manifest to upload.
 */
function processActions(
  iface: ts.InterfaceDeclaration,
  checker: ts.TypeChecker,
): BlockManifest['actions'] {
  return processInterface(iface, checker, (name, description) => [name ?? '$any', { description }]);
}

/**
 * Get an events object based on TypeScript interface nodes.
 *
 * @param eventListenerInterface The node to base the event listeners on.
 * @param eventEmitterInterface The node to base the event emitters on.
 * @param checker The TypeScript type checker.
 * @returns The events manifest to upload.
 */
function processEvents(
  eventListenerInterface: ts.InterfaceDeclaration,
  eventEmitterInterface: ts.InterfaceDeclaration,
  checker: ts.TypeChecker,
): BlockManifest['events'] {
  const listen = processInterface(eventListenerInterface, checker, (name, description) => [
    name ?? '$any',
    { description },
  ]);
  const emit = processInterface(eventEmitterInterface, checker, (name, description) => [
    name ?? '$any',
    { description },
  ]);

  return (emit || listen) && { emit, listen };
}

/**
 * Get the JSON schema for parameters based on a TypeScript program.
 *
 * @param program The TypeScript program from which to extract parameters.
 * @param iface The interface node from which to extract parameters.
 * @returns The JSON schema for the block parameters.
 */
function processParameters(program: ts.Program, iface: ts.InterfaceDeclaration): Schema {
  if (!iface) {
    return;
  }
  const formatter = createFormatter({}, (fmt) => {
    fmt.addTypeFormatter({
      supportsType: (type) => type.getName() === 'IconName',
      getDefinition: () => ({ type: 'string', format: 'fontawesome' }),
      getChildren: () => [],
    });
    fmt.addTypeFormatter({
      supportsType: (type) => type.getName() === 'Remapper',
      getDefinition: () => ({ format: 'remapper' }),
      getChildren: () => [],
    });
  });
  const parser = createParser(program, { expose: 'all', topRef: false });
  const generator = new SchemaGenerator(program, parser, formatter);
  const schema = generator.createSchemaFromNodes([iface]) as Schema;
  if (schema.definitions && !Object.keys(schema.definitions).length) {
    delete schema.definitions;
  }
  // This is the tsdoc that has been added to the SDK to aid the block developer.
  delete schema.description;
  return schema;
}

/**
 * Get a messages object based on a TypeScript interface node.
 *
 * @param iface The node to base the messages on.
 * @param checker The TypeScript type checker.
 * @returns The action manifest to upload.
 */
function processMessages(
  iface: ts.InterfaceDeclaration,
  checker: ts.TypeChecker,
): BlockManifest['messages'] {
  return processInterface(iface, checker, (name, description) => [name, { description }]);
}

/**
 * Get the TypeScript program for a given path.
 *
 * @param blockPath The path for which to get the TypeScript program.
 * @returns The TypeScript program.
 */
function getProgram(blockPath: string): ts.Program {
  const diagnosticHost: ts.FormatDiagnosticsHost = {
    getNewLine: () => ts.sys.newLine,
    getCurrentDirectory: ts.sys.getCurrentDirectory,
    getCanonicalFileName: (x) => x,
  };
  const tsConfigPath = ts.findConfigFile(blockPath, ts.sys.fileExists);
  const { config, error } = ts.readConfigFile(tsConfigPath, ts.sys.readFile);
  if (error) {
    throw new AppsembleError(ts.formatDiagnostic(error, diagnosticHost));
  }
  if (!config.files || !config.include) {
    config.files = ts.sys
      .readDirectory(blockPath, ['.ts', '.tsx'])
      .map((f) => relative(blockPath, f));
  }
  const { errors, fileNames, options } = ts.parseJsonConfigFileContent(
    config,
    ts.sys,
    blockPath,
    undefined,
    tsConfigPath,
  );
  // Filter: 'rootDir' is expected to contain all source files.
  const diagnostics = errors.filter(({ code }) => code !== 6059);
  if (diagnostics.length) {
    throw new AppsembleError(ts.formatDiagnosticsWithColorAndContext(diagnostics, diagnosticHost));
  }

  options.noEmit = true;
  delete options.out;
  delete options.outDir;
  delete options.outFile;
  delete options.declaration;
  delete options.declarationDir;
  delete options.declarationMap;
  const program = ts.createProgram(fileNames, options);
  const preEmitDiagnostics = ts.getPreEmitDiagnostics(program);
  if (preEmitDiagnostics.length) {
    throw new AppsembleError(
      ts.formatDiagnosticsWithColorAndContext(preEmitDiagnostics, diagnosticHost),
    );
  }
  return program;
}

/**
 * Generate a block manifest from the block metadata and TypeScript project.
 *
 * Uses the .appsemblerc file and the type definitions of the block.
 *
 * @param blockConfig The block configuration
 * @returns The block configuration appended from the TypeScript project.
 */
export function getBlockConfigFromTypeScript(
  blockConfig: BlockConfig,
): Pick<BlockManifest, 'actions' | 'events' | 'messages' | 'parameters'> {
  if ('actions' in blockConfig && 'events' in blockConfig && 'parameters' in blockConfig) {
    return blockConfig;
  }
  logger.info(`Extracting data from TypeScript project ${blockConfig.dir}`);
  const program = getProgram(blockConfig.dir);
  const checker = program.getTypeChecker();

  let actionInterface: ts.InterfaceDeclaration;
  let eventEmitterInterface: ts.InterfaceDeclaration;
  let eventListenerInterface: ts.InterfaceDeclaration;
  let messagesInterface: ts.InterfaceDeclaration;
  let patametersInterface: ts.InterfaceDeclaration;

  for (const sourceFile of program.getSourceFiles()) {
    const fileName = relative(process.cwd(), sourceFile.fileName);
    // Filter TypeScript default libs
    if (program.isSourceFileDefaultLibrary(sourceFile)) {
      logger.silly(`Skipping metadata extraction from: ${fileName}`);
      continue;
    }
    logger.verbose(`Searching metadata in: ${fileName}`);
    // eslint-disable-next-line @typescript-eslint/no-loop-func
    ts.forEachChild(sourceFile, (mod) => {
      // This node doesn’t override SDK types
      if (!ts.isModuleDeclaration(mod)) {
        return;
      }
      // This module defines other types
      if (mod.name.text !== '@appsemble/sdk') {
        return;
      }
      ts.forEachChild(mod.body, (iface) => {
        // Appsemble only uses module interface augmentation.
        if (!ts.isInterfaceDeclaration(iface)) {
          return;
        }
        const { line } = sourceFile.getLineAndCharacterOfPosition(iface.getStart(sourceFile));
        // Line numbers are 0 indexed, whereas they are usually represented as 1 indexed.
        const loc = `${normalizePath(fileName)}:${line + 1}`;

        switch (iface.name.text) {
          case 'Actions':
            logger.info(`Found augmented interface 'Actions' in '${loc}'`);
            if (actionInterface) {
              throw new AppsembleError(`Found duplicate interface 'Actions' in '${loc}'`);
            }
            actionInterface = iface;
            break;
          case 'EventEmitters':
            logger.info(`Found augmented interface 'EventEmitters' in '${loc}'`);
            if (eventEmitterInterface) {
              throw new AppsembleError(`Found duplicate interface 'EventEmitters' in '${loc}'`);
            }
            eventEmitterInterface = iface;
            break;
          case 'EventListeners':
            logger.info(`Found augmented interface 'EventListeners' in '${loc}'`);
            if (eventListenerInterface) {
              throw new AppsembleError(`Found duplicate interface 'EventListeners' in '${loc}'`);
            }
            eventListenerInterface = iface;
            break;
          case 'Parameters':
            logger.info(`Found augmented interface 'Parameters' in '${loc}'`);
            if (patametersInterface) {
              throw new AppsembleError(`Found duplicate interface 'Parameters' in '${loc}'`);
            }
            patametersInterface = iface;
            break;
          case 'Messages':
            logger.info(`Found augmented interface 'Messages' in '${loc}'`);
            messagesInterface = iface;
            break;
          default:
            logger.warn(`Detected unused augmented type ${iface.name.text} in ${loc}`);
        }
      });
    });
  }

  return {
    actions:
      'actions' in blockConfig ? blockConfig.actions : processActions(actionInterface, checker),
    events:
      'events' in blockConfig
        ? blockConfig.events
        : processEvents(eventListenerInterface, eventEmitterInterface, checker),
    parameters:
      'parameters' in blockConfig
        ? blockConfig.parameters
        : processParameters(program, patametersInterface),
    messages:
      'messages' in blockConfig
        ? blockConfig.messages
        : processMessages(messagesInterface, checker),
  };
}