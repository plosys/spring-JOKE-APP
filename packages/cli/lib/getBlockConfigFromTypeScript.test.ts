import { readdirSync } from 'node:fs';
import process from 'node:process';
import { fileURLToPath } from 'node:url';

import { AppsembleError, resolveFixture } from '@appsemble/node-utils';
import { ts } from 'ts-json-schema-generator';

import { getBlockConfigFromTypeScript } from './getBlockConfigFromTypeScript.js';

beforeEach(() => {
  import.meta.jest.spyOn(process, 'cwd').mockReturnValue(resolveFixture('.'));
});

describe('getBlockConfigFromTypeScript', () => {
  it('should extract configuration from a TypeScript project', () => {
    const result = getBlockConfigFromTypeScript({
      name: '',
      layout: 'float',
      version: '1.33.7',
      webpack: '',
      output: '',
      dir: resolveFixture('getBlockConfigFromTypeScript/valid'),
    });
    expect(result).toStrictEqual({
      actions: { testAction: { description: undefined } },
      events: {
        emit: { testEmit: { description: undefined } },
        listen: { testListener: { description: undefined } },
      },
      messages: undefined,
      parameters: {
        $schema: 'http://json-schema.org/draft-07/schema#',
        additionalProperties: false,
        properties: {
          param1: { type: 'string' },
          param2: { type: 'number' },
          param3: { type: 'boolean' },
          param4: { type: 'string' },
          param5: {
            additionalProperties: false,
            properties: {
              nested1: { type: 'string' },
              nested2: { type: 'number' },
              nested3: { type: 'boolean' },
            },
            required: ['nested1', 'nested2', 'nested3'],
            type: 'object',
          },
          param6: { items: { type: 'string' }, type: 'array' },
        },
        required: ['param1', 'param2', 'param3', 'param5', 'param6'],
        type: 'object',
      },
    });
  });

  it('should not use TypeScript if all metadata is present in the original config', () => {
    import.meta.jest.spyOn(ts, 'createProgram');
    const input = {
      actions: {},
      events: { emit: { foo: {} }, listen: { bar: {} } },
      layout: 'float',
      parameters: { type: 'object' },
      version: '1.33.7',
      webpack: '',
      output: '',
      dir: resolveFixture('getBlockConfigFromTypeScript/valid'),
      name: '',
    } as const;
    const result = getBlockConfigFromTypeScript(input);

    expect(result).toBe(input);
    expect(ts.createProgram).not.toHaveBeenCalled();
  });

  it('should prefer actions overrides over TypeScript actions', () => {
    const result = getBlockConfigFromTypeScript({
      actions: {},
      version: '1.33.7',
      webpack: '',
      output: '',
      dir: resolveFixture('getBlockConfigFromTypeScript/valid'),
      name: '',
    });

    expect(result.actions).toStrictEqual({});
  });

  it('should prefer events overrides over TypeScript events', () => {
    const result = getBlockConfigFromTypeScript({
      events: { emit: { onSuccess: {} } },
      version: '1.33.7',
      webpack: '',
      output: '',
      dir: resolveFixture('getBlockConfigFromTypeScript/valid'),
      name: '',
    });

    expect(result.events).toStrictEqual({ emit: { onSuccess: {} } });
  });

  it('should prefer parameters overrides over TypeScript parameters', () => {
    const result = getBlockConfigFromTypeScript({
      parameters: { type: 'object' },
      version: '1.33.7',
      webpack: '',
      output: '',
      dir: resolveFixture('getBlockConfigFromTypeScript/valid'),
      name: '',
    });

    expect(result.parameters).toStrictEqual({ type: 'object' });
  });

  it('should throw if duplicate Actions are found', () => {
    expect(() =>
      getBlockConfigFromTypeScript({
        webpack: '',
        output: '',
        dir: resolveFixture('getBlockConfigFromTypeScript/duplicateActions'),
        name: '',
        version: '1.33.7',
      }),
    ).toThrow(
      new AppsembleError(
        "Found duplicate interface 'Actions' in 'getBlockConfigFromTypeScript/duplicateActions/index.ts:31'",
      ),
    );
  });

  it('should throw if duplicate EventEmitters are found', () => {
    expect(() =>
      getBlockConfigFromTypeScript({
        webpack: '',
        output: '',
        dir: resolveFixture('getBlockConfigFromTypeScript/duplicateEventEmitters'),
        name: '',
        version: '1.33.7',
      }),
    ).toThrow(
      new AppsembleError(
        "Found duplicate interface 'EventEmitters' in 'getBlockConfigFromTypeScript/duplicateEventEmitters/index.ts:31'",
      ),
    );
  });

  it('should throw if duplicate EventListeners are found', () => {
    expect(() =>
      getBlockConfigFromTypeScript({
        webpack: '',
        output: '',
        dir: resolveFixture('getBlockConfigFromTypeScript/duplicateEventListeners'),
        name: '',
        version: '1.33.7',
      }),
    ).toThrow(
      new AppsembleError(
        "Found duplicate interface 'EventListeners' in 'getBlockConfigFromTypeScript/duplicateEventListeners/index.ts:31'",
      ),
    );
  });

  it('should throw if duplicate Parameters are found', () => {
    expect(() =>
      getBlockConfigFromTypeScript({
        web