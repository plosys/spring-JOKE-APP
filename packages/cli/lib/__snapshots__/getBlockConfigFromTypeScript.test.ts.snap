// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBlockConfigFromTypeScript official blocks action-button 1`] = `
{
  "actions": {
    "onClick": {
      "description": "Action that gets dispatched when the button is clicked.",
    },
  },
  "events": undefined,
  "messages": undefined,
  "parameters": {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "additionalProperties": false,
    "properties": {
      "icon": {
        "description": "A [Font Awesome icon](https://fontawesome.com/icons?m=free) name to render on the button.",
        "format": "fontawesome",
        "type": "string",
      },
      "title": {
        "description": "The title for the button.

Describe what the button does. This helps with accessibility for people using screen readers.",
        "format": "remapper",
      },
    },
    "required": [
      "icon",
    ],
    "type": "object",
  },
}
`;

exports[`getBlockConfigFromTypeScript official blocks audio 1`] = `
{
  "actions": undefined,
  "events": {
    "emit": undefined,
    "listen": {
      "onAudio": {
        "description": "The event to listen on for new data.

The result must be either a audio URL or an object containing a property \`url\`.",
      },
      "stop": {
        "description": "The event to stop listening to audio.",
      },
    },
  },
  "messages": undefined,
  "parameters": {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "additionalProperties": false,
    "properties": {
      "src": {
        "description": "The source of the player. This can either be a URL pointing to a website, or it can point to an asset. The asset can either be the object, or the ID of the asset",
        "format": "remapper",
      },
    },
    "type": "object",
  },
}
`;

exports[`getBlockConfigFromTypeScript official blocks button-list 1`] = `
{
  "actions": {
    "$any": {
      "description": "A custom action that gets dispatched when a button is clicked that has the same click action
specified as the name of this action.",
    },
    "onClick": {
      "description": "Action that gets dispatched when a button is clicked that doesn’t specify its own click
action.",
    },
  },
  "events": {
    "emit": undefined,
    "listen": {
      "data": {
        "description": "The event that is triggered when data is received.

This data can be used with remap to display labels dynamically based on the received data.",
      },
    },
  },
  "messages": undefined,
  "parameters": {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "additionalProperties": false,
    "definitions": {
      "BulmaColor": {
        "description": "A color known to Bulma.",
        "enum": [
          "danger",
          "dark",
          "info",
          "link",
          "primary",
          "success",
          "warning",
          "white",
        ],
        "type": "string",
      },
      "BulmaSize": {
        "enum": [
          "large",
          "medium",
          "normal",
          "small",
        ],
        "type": "string",
      },
      "Button": {
        "additionalProperties": false,
        "properties": {
          "color": {
            "$ref": "#/definitions/BulmaColor",
            "description": "The Bulma color to use.",
          },
          "fullwidth": {
            "description": "Whether the button should be full width or not.

By default buttons only take up as much space as needed.",
            "type": "boolean",
          },
          "icon": {
            "description": "A [Font Awesome icon](https://fontawesome.com/icons?m=free) name to render on the button.

Will not render if undefined.",
            "format": "fontawesome",
            "type": "string",
          },
          "inverted": {
            "description": "Whether the text and background colors should be inverted.",
            "type": "boolean",
          },
          "label": {
            "description": "The label to display.

Will not render if undefined.",
            "format": "remapper",
          },
          "light": {
            "description": "When set to true, the ‘light’ set of Bulma colors are used.",
            "type": "boolean",
          },
          "onClick": {
            "description": "The name of the action to trigger when the button is clicked.",
            "format": "action",
            "type": "string",
          },
          "outlined": {
            "description": "Whether the button should display its colors in the outlines.",
            "type": "boolean",
          },
          "rounded": {
            "description": "Whether the button should be rounded.",
            "type": "boolean",
          },
          "size": {
            "$ref": "#/definitions/BulmaSize",
            "description": "The size of the button. By default the size is ‘normal’.",
          },
          "title": {
            "description": "The title for the button.

Describe what the button does. This helps with accessibility for people using screen readers.",
            "format": "remapper",
          },
        },
        "type": "object",
      },
    },
    "properties": {
      "buttons": {
        "description": "The list of buttons.",
        "items": {
          "$ref": "#/definitions/Button",
        },
        "type": "array",
      },
    },
    "required": [
      "buttons",
    ],
    "type": "object",
  },
}
`;

exports[`getBlockConfigFromTypeScript official blocks chart 1`] = `
{
  "actions": {
    "onClick": {
      "description": "This action is launched whenever a user clicks on the chart canvas.

Returns the data of the clicked data in the format of \`label\`, \`data\`.

Returns an empty object if the user clicks outside of a data entry.",
    },
  },
  "events": {
    "emit": undefined,
    "listen": {
      "add": {
        "description": "Add new data to the chart.

Refer to the block description for more information on the supported data format.",
      },
      "replace": {
        "description": "Replace all current data with the new data.

Refer to the block description for more information on the supported data format.",
      },
    },
  },
  "messages": undefined,
  "parameters": {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "additionalProperties": false,
    "definitions": {
      "Color": {
        "description": "A hexadecimal color.",
        "pattern": "^#[a-fA-F\\d]{6}$",
        "type": "string",
      },
      "Font": {
        "additionalProperties": false,
        "properties": {
          "size": {
            "description": "The size of the label's font in px.",
            "type": "number",
          },
        },
        "required": [
          "size",
        ],
        "type": "object",
      },
      "LabelOptions": {
        "additionalProperties": false,
        "properties": {
          "font": {
            "$ref": "#/definitions/Font",
            "description": "Options for the font style for the labels on the horizontal axis.",
          },
          "maxWidth": {
            "description": "The maximum length (words) of the labels on the horizontal axis.",
            "type": "number",
          },
        },
        "required": [
          "font",
          "maxWidth",
        ],
        "type": "object",
      },
      "YAxis": {
        "additionalProperties": false,
        "properties": {
          "colors": {
            "default": [
              "#ededed",
            ],
            "description": "By default a grey tone is used.",
            "items": {
              "$ref": "#/definitions/Color",
            },
            "minItems": 1,
            "type": "array",
          },
          "max": {
            "description": "The maximum value to render on the vertical axis.

If not specified, the maximum value is determined automatically based on the values given.",
            "type": "number",
          },
          "min": {
            "default": 0,
            "description": "The minimal value to render on the vertical axis.",
            "type": "number",
          },
          "step": {
            "description": "If specified, the Y-axis of the chart will be incremented by this fixed amount.

By default it’s detected automatically.",
            "type": "number",
          },
        },
        "type": "object",
      },
    },
    "properties": {
      "backgroundColors": {
        "description": "The background colors to use for each dataset by default.

If the length of this array exceeds the number of labels, the pattern will be repeated.

By default the primary color of the theme will be used.",
        "items": {
          "format": "remapper",
        },
        "minItems": 1,
        "type": "array",
      },
      "labelOptions": {
        "$ref": "#/definitions/LabelOptions",
        "description": "Configuration options for the labels on the horizontal axis.",
      },
      "labels": {
        "description": "The labels to display at the bottom of the chart.",
        "items": {
          "format": "remapper",
        },
        "minItems": 1,
        "type": "array",
      },
      "type": {
        "default": "line",
        "description": "The chart type to render by default.",
        "enum": [
          "bar",
          "line",
        ],
        "type": "string",
      },
      "yAxis": {
        "$ref": "#/definitions/YAxis",
        "description": "Configuration options for the vertical axis.",
      },
    },
    "required": [
      "type",
      "yAxis",
    ],
    "type": "object",
  },
}
`;

exports[`getBlockConfigFromTypeScript official blocks control-buttons 1`] = `
{
  "actions": {
    "onBack": {
      "description": "This action gets triggered then the back button is clicked.

It will be called with the page data.",
    },
    "onForward": {
      "description": "This action gets triggered then the back button is clicked.

It will be called with the page data.",
    },
  },
  "events": undefined,
  "messages": {
    "back": {
      "description": "The label that’s used on the back button.",
    },
    "forward": {
      "description": "The label that’s used on the forward button.",
    },
  },
  "parameters": {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "additionalProperties": false,
    "definitions": {
      "Button": {
        "additionalProperties": false,
        "properties": {
          "icon": {
            "description": "A [Font Awesome icon](https://fontawesome.com/icons?m=free) name to render on the button.

By default a cared pointing left or right respectively will be rendered.",
            "format": "fontawesome",
            "type": "string",
          },
        },
        "type": "object",
      },
    },
    "properties": {
      "back": {
        "anyOf": [
          {
            "$ref": "#/definitions/Button",
          },
          {
            "const": false,
            "type": "boolean",
          },
        ],
        "description": "The configuration for the back button.

Set to \`false\` to hide back button.",
      },
      "forward": {
        "anyOf": [
          {
            "$ref": "#/definitions/Button",
          },
          {
            "const": false,
            "type": "boolean",
          },
        ],
        "description": "The configuration for the forward button.

Set to \`false\` to hide forward button.",
      },
    },
    "type": "object",
  },
}
`;

exports[`getBlockConfigFromTypeScript official blocks data-loader 1`] = `
{
  "actions": {
    "onLoad": {
      "description": "Action that gets dispatched when a new filter gets applied.

This also gets called during the initial load.",
    },
  },
  "events": {
    "emit": {
      "data": {
        "description": "Event that gets emitted once the \`onLoad\` action has finished.

It can be triggered again by sending a \`refresh\` event.",
      },
    },
    "listen": {
      "refresh": {
        "description": "When received, the \`onLoad\` action will be triggered using the parameters passed through this
event, which in turn triggers the \`data\` emit event.",
      },
    },
  },
  "messages": {
    "loadErrorMessage": {
      "description": "The error message that should be shown when data failed to load.",
    },
  },
  "parameters": {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "additionalProperties": false,
    "properties": {
      "skipInitialLoad": {
        "description": "By default the \`onLoad\` action is triggered immediately.

By setting this to \`true\`, this won’t happen.",
        "type": "boolean",
      },
    },
    "type": "object",
  },
}
`;

exports[`getBlockConfigFromTypeScript official blocks data-notifier 1`] = `
{
  "actions": undefined,
  "events": {
    "emit": {
      "data": {
        "description": "Event that gets emitted when new data is available.",
      },
    },
    "listen": {
      "data": {
        "description": "The event to listen on for new data.",
      },
      "seed": {
        "description": "Set and emit the initial data.

This is useful for example in combination with the \`filter\` block.",
      },
    },
  },
  "messages": {
    "buttonLabel": {
      "description": "The label to display on the refresh button of the notification.",
    },
    "newMessage": {
      "description": "The message to display when new data is available.

This will be called with the \`count\` parameter, which refers to the amount of new items.",
    },
    "updatedMessage": {
      "description": "The message to display when existing data has been changed.

This will be called with the \`count\` parameter, which refers to the amount of changed items.",
    },
  },
  "parameters": {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "additionalProperties": false,
    "definitions": {
      "BulmaColor": {
        "description": "A color known to Bulma.",
        "enum": [
          "danger",
          "dark",
          "info",
          "link",
          "primary",
          "success",
          "warning",
          "white",
        ],
        "type": "string",
      },
    },
    "properties": {
      "color": {
        "$ref": "#/definitions/BulmaColor",
        "default": "dark",
        "description": "The Bulma color to use for the message and the refresh button.",
      },
      "id": {
        "default": "id",
        "description": "The property used to identify resources when comparing them in case of array data.",
        "type": "string",
      },
    },
    "type": "object",
  },
}
`;

exports[`getBlockConfigFromTypeScript official blocks detail-viewer 1`] = `
{
  "actions": undefined,
  "events": {
    "emit": undefined,
    "listen": {
      "data": {
        "description": "The event that is triggered when data is received.

Compatible data that is received will be displayed.",
      },
    },
  },
  "messages": undefined,
  "parameters": {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "additionalProperties": false,
    "definitions": {
      "AssetMarkerIcon": {
        "additionalProperties": false,
        "description": "A marker based on an existing asset.",
        "properties": {
          "anchor": {
            "description": "The anchor X and Y offset used for positioning the image.

By default, the center of the icon will be used to mark the location. For many icons, it may be desirable to customize this. For example, for a symmetric pin which has a width of 10, and a height of 16, you’ll probably want to set this to \`[5, 16]\`

The following special cases for [Font Awesome icon](https://fontawesome.com/icons?m=free) are treated in a special way, since they are often used to represent a location:

- \`map-marker\`
- \`map-marker-alt\`
- \`map-pin\`
- \`thumbtrack\`",
            "items": {
              "type": "number",
            },
            "maxItems": 2,
            "minItems": 2,
            "type": "array",
          },
          "asset": {
            "description": "The id of an asset to use.",
            "type": "string",
          },
          "size": {
            "default": 28,
            "description": "The height of marker icons in pixels.",
            "type": "number",
          },
        },
        "required": [
          "asset",
        ],
        "type": "object",
      },
      "BulmaColor": {
        "description": "A color known to Bulma.",
        "enum": [
          "danger",
          "dark",
          "info",
          "link",
          "primary",
          "success",
          "warning",
          "white",
        ],
        "type": "string",
      },
      "Field": {
        "anyOf": [
          {
            "$ref": "#/definitions/FileField",
          },
          {
            "$ref": "#/definitions/GeoCoordinatesField",
          },
          {
            "$ref": "#/definitions/StringField",
          },
        ],
        "description": "All supported types of fields.",
      },
      "FieldGroup": {
        "additionalProperties": false,
        "description": "A group of fields that is repeated for each item in it value.",
        "properties": {
          "fields": {
            "description": "The list of fields to repeat.",
            "items": {
              "$ref": "#/definitions/Field",
            },
            "type": "array",
          },
          "label": {
            "description": "The label that is presented to the user. No label will be displayed if this is not defined.",
            "format": "remapper",
          },
          "value": {
            "description": "The Remapper used to retrieve the data.",
            "format": "remapper",
          },
        },
        "required": [
          "fields",
        ],
        "type": "object",
      },
      "FileField": {
        "additionalProperties": false,
        "description": "Displays files as images.",
        "properties": {
          "label": {
            "description": "The label that is presented to the user. No label will be displayed if this is not defined.",
            "format": "remapper",
          },
          